### PR TITLE
Fix endianness in various files

### PR DESCRIPTION
**Pull request by wargio**
# do not squash

 

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/bin/p/bin_mbn.c [hunk 1]
```diff
@@ -22,105 +21,102 @@ typedef struct sbl_header {
...
-		// TODO: Add more checks here
-		return true;
-	}
-	return false;
+	return read_sbl_header(b, &sb);
 }
 
 static bool load_buffer(RzBinFile *bf, RzBinObject *obj, RzBuffer *b, Sdb *sdb) {
-	return check_buffer(b);
+	SblHeader *sb = RZ_NEW0(SblHeader);
```
Comment (XVilka) [inline librz/bin/p/bin_mbn.c:73]: Check for NULL

File: librz/bin/p/bin_mbn.c [hunk 2]
```diff
@@ -22,105 +21,102 @@ typedef struct sbl_header {
 	ut32 cert_sz;
 } SblHeader;
 
-// TODO avoid globals
-static SblHeader sb = { 0 };
+static bool read_sbl_header(RzBuffer *b, SblHeader *sb) {
+	ut64 offset = 0;
+	if (!(rz_buf_read_le32_offset(b, &offset, &sb->load_index) &&
```
Comment (XVilka) [inline librz/bin/p/bin_mbn.c]: Probably better to split this into two functions - one for reading like `read_sbl_header` and one for checking.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/bin/format/le/le.c [hunk 3]
```diff
@@ -94,6 +94,28 @@ static RzBinSymbol *le_get_symbol(rz_bin_le_obj_t *bin, ut64 *offset) {
 	return sym;
 }
 
+static bool read_le_entry_bundle_entry(RzBuffer *buf, ut64 offset, LE_entry_bundle_entry *e, LE_entry_bundle_type type) {
+	switch (type) {
+	case ENTRY16:
+		return rz_buf_read8_offset(buf, &offset, &e->entry_16.flags) &&
```
Comment (XVilka) [inline librz/bin/format/le/le.c:101]: I am not sure about taking the address of the argument variable. What if it becomes a register on some architectures?

File: librz/bin/p/bin_mbn.c [hunk 4]
```diff
@@ -22,105 +21,102 @@ typedef struct sbl_header {
...
-		// TODO: Add more checks here
-		return true;
+	if (!read_sbl_header(b, &sb)) {
+		return false;
 	}
-	return false;
+	if (sb.version != 3) { // NAND
+		return false;
+	}
+	if (sb.paddr + sizeof(SblHeader) > offset) { // NAND
```
Comment (XVilka) [inline librz/bin/p/bin_mbn.c:47]: `offset` is undefined here:
```
 tcc -Ilibrz/bin/librz_bin.so.0.5.0.p -I. -I.. -Ilibrz -I../librz -Ilibrz/include -I../librz/include -I../librz/bin/format -I../subprojects/lz4-1.9.3/lib -Isubprojects/yxml -I../subprojects/yxml -Ilibrz/util/sdb/src -I../librz/util/sdb/src -Ilibrz/util/sdb -I../librz/util/sdb -I../librz/type/parser -I../subprojects/libmspack/libmspack/mspack -I../subprojects/libmspack/libmspack -Wall -O2 -g -Wimplicit-fallthrough=3 -DRZ_PLUGIN_INCORE=1 -D_GNU_SOURCE -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DRZ_API_BIN_ONLY=1 -MD -MF librz/bin/librz_bin.so.0.5.0.p/p_bin_mbn.c.o.d -o librz/bin/librz_bin.so.0.5.0.p/p_bin_mbn.c.o -c ../librz/bin/p/bin_mbn.c
../librz/bin/p/bin_mbn.c:47: error: 'offset' undeclared
```

File: librz/core/cmd/cmd_write.c [hunk 5]
```diff
@@ -76,15 +76,21 @@ static bool encrypt_or_decrypt_block(RzCore *core, const char *algo, const char
 }
 
 static void cmd_write_bits(RzCore *core, int set, ut64 val) {
+	ut8 buf[sizeof(ut64)];
 	ut64 ret, orig;
 	// used to set/unset bit in current address
-	rz_io_read_at(core->io, core->offset, (ut8 *)&orig, sizeof(orig));
```
Comment (XVilka) [inline librz/core/cmd/cmd_write.c:81]: Nice finding

File: librz/util/mem.c [hunk 6]
```diff
@@ -186,17 +186,6 @@ RZ_API int rz_mem_set_num(ut8 *dest, int dest_size, ut64 num) {
...
-	if (big_endian) {
-		rz_mem_swapendian(dest, src, len);
-	} else {
-		memcpy(dest, src, len);
-	}
-}
-
 // This function unconditionally swaps endian of size bytes of orig -> dest
 // TODO: Remove completely
 RZ_API void rz_mem_swapendian(ut8 *dest, const ut8 *orig, int size) {
```
Comment (XVilka) [inline librz/util/mem.c:191]: Maybe remove this one too?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/bin/format/le/le.c (hunk 3 above)
Comment (wargio) [reply librz/bin/format/le/le.c:101]: good point.

File: librz/util/mem.c (hunk 6 above)
Comment (wargio) [reply librz/util/mem.c:191]: i'm trying too, but it is used in various places still.

File: librz/bin/format/le/le.c (hunk 3 above)
Comment (XVilka) [reply librz/bin/format/le/le.c:101]: Don't forget this one please

**Review by XVilka — APPROVED** [review]
LGTM once comment is addressed and everything is green.

**Pull request merged**